File: source/downloads/code/LeetCode/Python/Group-Anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """

        ret = []
        # hash_map: <strs[i], 和strs[i]互为anagrams的其他字符串>
        hash_map = dict()
        for i , s in enumerate(strs):
            target = "".join(sorted(s))
            if target not in hash_map:
                hash_map[target] = [s]
            else:
                hash_map[target].append(s)
        
        for l in hash_map.values():
            # l 就是所有anagrams的列表
            ret.append(sorted(l))
        
        return ret

File: source/downloads/code/LeetCode/Java/Group-Anagrams.java
public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if(strs==null || strs.length==0) return new ArrayList<List<String>>();
        Map<String, List<String>> map = new HashMap<String, List<String>>();
        Arrays.sort(strs);
        for(String s : strs){
            char[] char_s = s.toCharArray();
            Arrays.sort(char_s);
            String key_val = String.valueOf(char_s);
            if(!map.containsKey(key_val))
                map.put(key_val, new ArrayList<String>());
            map.get(key_val).add(s);
        }
        return new ArrayList<List<String>>(map.values());
    }
}
